src/irmin-pack/inode.ml

```diff
@@ -327,24 +327,65 @@ struct
               t.tree <- Some x;
               x)
 
-    let rec list_entry ~find acc = function
-      | Empty -> acc
-      | Inode i -> list_values ~find acc (get_tree ~find i)
+    let rec drop n (l : 'a Seq.t) () =
+      match l () with
+      | l' when n = 0 -> l'
+      | Nil -> Nil
+      | Cons (_, l') -> drop (n - 1) l' ()
+
+    let rec take n (l : 'a Seq.t) () =
+      if n = 0 then Seq.Nil
+      else
+        match l () with Nil -> Nil | Cons (x, l') -> Cons (x, take (n - 1) l')
 
-    and list_inodes ~find acc t =
-      Array.fold_left (list_entry ~find) acc t.entries
+    type acc = { cursor : int; values : (step * value) list; remaining : int }
 
-    and list_values ~find acc t =
-      match t.v with
-      | Values vs -> StepMap.bindings vs @ acc
-      | Inodes t -> list_inodes ~find acc t
+    let empty_acc n = { cursor = 0; values = []; remaining = n }
+
+    let rec list_entry ~offset ~length ~find acc = function
+      | Empty -> acc
+      | Inode i -> list_values ~offset ~length ~find acc (get_tree ~find i)
 
-    let compare_step = Irmin.Type.(unstage (compare step_t))
-    let compare_entry x y = compare_step (fst x) (fst y)
+    and list_inodes ~offset ~length ~find acc t =
+      if acc.remaining <= 0 || offset + length < acc.cursor then acc
+      else if acc.cursor + Array.length t.entries < offset then
+        { acc with cursor = Array.length t.entries + acc.cursor }
+      else Array.fold_left (list_entry ~offset ~length ~find) acc t.entries
 
-    let list ~find t =
-      let entries = list_values ~find [] t in
-      List.fast_sort compare_entry entries
+    and list_values ~offset ~length ~find acc t =
+      if acc.remaining <= 0 || offset + length < acc.cursor then acc
```

Comment:
```suggestion
      if acc.remaining <= 0 || offset + length <= acc.cursor then acc
```

I think we're done if `offset + length = acc.cursor`, since we should have read `length` elements.